### PR TITLE
tests: translated log-analyzer script from bash to python

### PR DESCRIPTION
**Pull request by maykathm**
This replaces the log-analyzer script from bash to python. The bash script, because of many calls to `spread -list` when getting the list of tasks to reexecute, could require upwards of 20 minutes to execute. The new Python script will call `spread -list` exactly once. This additionally adds some python unit tests to ensure the script's correctness.

### CODE REVIEW

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 1]
```diff
@@ -1,330 +1,156 @@
...
+#!/usr/bin/env python3
+
+import argparse
+import json
+import os
+import subprocess
+import sys
+
+
+def filter_with_spread(exec_param: str):
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:10]: Just guessing:

```suggestion
def filter_with_spread(exec_param: str) -> list[str]:
```

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 2]
```diff
@@ -1,330 +1,156 @@
...
+    exec_param = exec_param.replace(',', ' ')
+    cmd = ['spread', '-list']
+    cmd.extend(exec_param.split())
+    return subprocess.check_output(cmd, universal_newlines=True).splitlines()
+
+
+def list_executed_tasks(filtered_exec_param: set, spread_logs: dict):
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:17]: Set of what? Dict of what?

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 3]
```diff
@@ -1,330 +1,156 @@
...
+    exec_and_failed = exec_and_failed.intersection(filtered_exec_param)
+    union = aborted_tasks.union(exec_and_failed)
+    if len(filtered_exec_param.difference(union)) == 0:
+        return set(exec_param.split())
+    return union
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description="""
+        usage:  log-analyzer list-failed-tasks <EXEC-PARAM> <PARSED-LOG>
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer]:
```suggestion
        Usage:  log-analyzer list-failed-tasks <EXEC-PARAM> <PARSED-LOG>
```

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 4]
```diff
@@ -1,330 +1,156 @@
...
+            exit(1)
+
+        with open(args.parsed_log, 'r') as f:
+            log = json.load(f)
+
+        if not log:
+            print("log.analyzer: the log file cannot be empty", file=sys.stderr)
+            exit(1)
+
+    if args.command == 'list-failed-tasks':
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:140]: You can also use sub-parsers and attach a function to a namespace produced by each parser to both handle `cmd --help`, `cmd subcmd --help` and to dispatch execution.

**Review by zyga — COMMENTED** [review]
Thank you for tackling this problem. This much welcome improvement.

I left a few quick comments but I'd really like to see type annotations and type-clean `mypy --strict` pass. Those little utilities tend to break from time to time and when one tries to reverse engineer what's going on it's often that the types fall apart in some edge cases - exactly what mypy can automatically help figure out.

I'm super happy to see unit tests, this is exactly what we should have always done.

I left some more comments on argument parsing details. I think you can really refactor that part and get very clean code with great user experience (`--help` and the like).

One word of caution, I'm not sure exactly if we are bound by python version requirements to run this (do we ever run this with ancient python?) I'd be really good to clearly document that in the log analyzer itself, as a longer code comment.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 2 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:17]: The only reason I hadn't added type hints with subscription or from the Typing module is to allow for execution with Python versions < 3.5. Type hints would obviously be ideal, but how about if I added documentation to each method instead? I could remove type hints altogether, but specify the type in the doc strings.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 1 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:10]: Using `list[str]` as a type hint requires Python 3.9 or higher. Would adding documentation to this function instead work for you? I could also leave list by itself so `def filter_with_spread(exec_param: str) -> list:`

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 4 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:140]: I looked through the docs for sub parsers and, unless I misread, the dispatch is only available from Python 3.13 on. I will modify with sub parsers though, thanks

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 1 above)
Comment (bboozzoo) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:10]: It's fine, I don't think we expect this to run on 3.9 anyway. The self hosted runners are likely much newer than 20.04?

Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer]: I updated to use subscription for type hints.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 4 above)
Comment (zyga) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:155]: I didn't knew about dispatch. I usually did this trick where each subparser had a default for "func" that was the code to run. At the end I called parse_arguments and then ns.func(...) if func was not None (if it was you didn't hit a sub-command).

Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:155]: After a few attempts, the most compact and comprehensible version I came up with is down below (the only other modification is adding `kwargs` to the necessary functions, but I didn't think it necessary to include here). In this particular case, I found this generated less code and was more readable than the monkey patching approach.

I think it's of lesser quality than the current version because it's roughly the same amount of code as the current version and it's not as straight forward. In this case, I think using `kwargs` to get away with passing more parameters than what the function calls for is a little hacky.

```python
def process_args(args):
    filtered_exec_param = set(filter_with_spread(args.exec_params))
    log = None
    if hasattr(args, 'parsed_log'):
        log = json.load(args.parsed_log)
        if not log:
            raise RuntimeError("log.analyzer: the log file cannot be empty")
    return {'exec_param':args.exec_params, 'filtered_exec_param':filtered_exec_param, 'spread_logs':log}


def setup_subparser(subparsers, function_dictionary, cmd, help, function) -> None:
    parser = subparsers.add_parser(cmd, help=help)
    add_arguments(parser)
    function_dictionary[cmd] = lambda **kwargs: print(' '.join(function(**kwargs)))


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="""
       ...
                                     """, formatter_class=argparse.RawDescriptionHelpFormatter)

    subparsers = parser.add_subparsers(dest='command')
    subparsers.required = True
    function_dictionary = {}
    setup_subparser(subparsers, function_dictionary, 'list-failed-tasks',
                    'list the tasks that failed during execute', list_failed_tasks)
    setup_subparser(subparsers, function_dictionary, 'list-executed-tasks',
                    'list the tasks that were executed', list_executed_tasks)
    setup_subparser(subparsers, function_dictionary, 'list-successful-tasks',
                    'list the successful tasks', list_successful_tasks)
    setup_subparser(subparsers, function_dictionary, 'list-aborted-tasks',
                    'list the aborted tasks (needs spread to be installed)', list_aborted_tasks)
    setup_subparser(subparsers, function_dictionary, 'list-reexecute-tasks',
                    'list the tasks to re-execute to complete (includes aborted and failed tasks)', list_rexecute_tasks)
    list_all = subparsers.add_parser(
        'list-all-tasks', help='list all the tasks')
    list_all.add_argument(
        'exec_params', help='This is the parameter used to run spread (something like this BACKEND:SYSTEM:SUITE)')
    function_dictionary['list-all-tasks'] = lambda **kwargs: ' '.join(print(kwargs['filtered_exec_param']))

    args = parser.parse_args()

    function_dictionary[args.command](**process_args(args))
```

Comment (zyga) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:155]: I think all the versions are fine. I meant for something much simpler but don't think that this has to be changed or is blocked by that aspect of my review.

**Comment by zyga** [comment]
@maykathm I've pushed a change that tries to illustrate the remaining type problem (and indeed a realization that the underlying data format is just weakly specified):

```
zyga@CFSR4-1:~/Projects/snapd/tests/lib/external/snapd-testing-tools/utils$ mypy --strict log-analyzer test_log_analyzer.py 
log-analyzer:53: error: TypedDict "SpreadLog_TypePhase" has no key "detail"  [typeddict-item]
test_log_analyzer.py:82: error: Missing key "stage" for TypedDict "SpreadLog_TypeResult"  [typeddict-item]
Found 2 errors in 2 files (checked 2 source files)
```

Have a look at the patches I've pushed to this branch please.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 5]
```diff
@@ -1,330 +1,220 @@
...
+    level: str
+    stage: str
+    detail: SpreadLogDetail
+
+
+SpreadLog = Union[SpreadLog_TypePhase, SpreadLog_TypeResult]
+
+
+def filter_with_spread(exec_param: str) -> list[str]:
+    exec_param = exec_param.replace(",", " ")
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:33]: Should we instead define the type of the argument as `string` with `nargs="+"` and remove this replacement and the split below?

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 6]
```diff
@@ -1,330 +1,220 @@
...
+    cmd = ["spread", "-list"]
+    cmd.extend(exec_param.split())
+    return subprocess.check_output(cmd, universal_newlines=True).splitlines()
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:36]: I think that `universal_newlines=` is deprecated and replaced by `text=`.

```suggestion
    return subprocess.check_output(cmd, text=True).splitlines()
```

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 7]
```diff
@@ -1,330 +1,220 @@
...
+        for log in spread_logs
+        if log["type"] == "phase" and log["verb"] == "Executing"
+    ]
+    return filtered_exec_param.intersection(executed)
+
+
+def _get_detail_lines(
+    spread_logs: list[SpreadLog], log_condition_func: Callable[[SpreadLog], bool]
+) -> list[str]:
+    result = [log["detail"]["lines"] for log in spread_logs if log_condition_func(log)]
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:53]: Do all the log "types" have a detail that has lines?

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 8]
```diff
@@ -1,330 +1,220 @@
...
+
+    def clean_entry(entry: str) -> str:
+        entry = entry.strip()
+        if entry.startswith("-"):
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:57]: Perhaps `entry.removeprefix` so that you can chain the whole thing to:
```python
def clean_line(line: str) -> str:
    return line.strip().removeprefix('-').strip()
```

Can you clarify why we strip the leading dash though? Ideally as a code comment.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 9]
```diff
@@ -1,330 +1,220 @@
...
+
+
+class SpreadLogDetail(TypedDict):
+    lines: list[str]
+
+
+class SpreadLog_TypeResult(TypedDict):
+    type: Literal["result"]
+    result_type: str
+    level: str
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:24]: I'd love to have a literal/enum that lists all known levels.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 10]
```diff
@@ -1,330 +1,220 @@
...
+    verb: str
+
+
+class SpreadLogDetail(TypedDict):
+    lines: list[str]
+
+
+class SpreadLog_TypeResult(TypedDict):
+    type: Literal["result"]
+    result_type: str
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:23]: I'd love to have a literal/enum that lists all known result types.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 11]
```diff
@@ -1,330 +1,220 @@
...
+    level: str
+    stage: str
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:25]: I'd love to have a literal/enum that lists all known stages.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer [hunk 12]
```diff
@@ -1,330 +1,220 @@
...
+    list_all = subparsers.add_parser("list-all-tasks", help="list all the tasks")
+
+    add_arguments(failed)
+    add_arguments(executed)
+    add_arguments(successful)
+    add_arguments(aborted)
+    add_arguments(reexecute)
+    list_all.add_argument(
+        "exec_params",
+        help="This is the parameter used to run spread (something like this BACKEND:SYSTEM:SUITE)",
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-analyzer:195]: Nitpick, this is copy-pasted from and shares a sibling above. We should clarify that the list is comma-separated or use `nargs='+'` if this does not break external format constraints.

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 9 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer]: Added as a comment

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 10 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer]: Added as a comment

File: tests/lib/external/snapd-testing-tools/utils/log-analyzer (hunk 11 above)
Comment (maykathm) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:26]: I believe the stage is read directly from the spread log so it is just the stages in spread: prepare, restore, execute

Comment (zyga) [reply tests/lib/external/snapd-testing-tools/utils/log-analyzer:26]: Thanks, I'll do a follow-up that tries to make this more explicit.

**Review by zyga — APPROVED** [review]
LGTM! Thank you!

I'd like to do a follow-up that moves some of the python types module as we will have more and more scripts that kind of do the same thing and I'd like to implement it fewer rather than more times.

**Comment by sergiocazzolato** [comment]
Thanks for this improvement. Also I think we need to clean the backends/systems we have in out spread.yaml because spread in order to list the jobs that match with the filter, calculates all the combinations and then apply the filter to all the jobs.

**Review by sergiocazzolato — APPROVED** [review]
thanks
